### ISSUE TITLE
Llama compiled artifacts are not properly reloaded

### ISSUE THREAD

**Issue by dacorvo** [aws-neuron/transformers-neuronx#39]
Looking at the source code of the latest package (`0.60.106`) it seems that the precompiled artifacts are not properly reloaded when `to_neuron()` is called.

To be more specific, it seems that the model contains a `main` head and several `context` heads.

Note: my understanding is that the `main` head is used to generate new tokens while the `context` heads are all in charge of the encoding of a subset of the context. Please correct me if I am wrong.

The precompiled artifacts are correctly reloaded for the `main` head, but not for the `context` heads, because the path to the precompiled artifacts is not passed down to them.

The following change seems to fix the issue:

`decoder.py`:

```
def build_weight_shared(self, n_positions_list=None, n_active_tokens=None, batch_size=None,
                       unroll=None, share_caches=False):
    ...
    new = DecoderLmHeadForSamplingNoEmbedding(
                self.tp_degree, n_positions_list, n_active_tokens, batch_size, self.attention_head_size,
                self.amp, self.num_layers, unroll, neuron_config=self.neuron_config, allow_pad=self.allow_pad
    )
+   new.compiler_artifacts_path = self.compiler_artifacts_path
    new.add_inputs_builder(self.inputs_builder)
```

**Comment by jeffhataws** [comment]
Thanks @dacorvo . We will take a look.

**Comment by shebbur-aws** [comment]
Hi @dacorvo, your understanding is correct, we have multiple NEFFs that needs to be serialized including token generation NEFFs and context encoding NEFFs. However they are not same NEFFs so we cannot simply do
new.compiler_artifacts_path = self.compiler_artifacts_path
There is an internal engineer working on a more general serialization API, which takes care of context encoding models as well.

**Comment by jeffhataws** [comment]
@dacorvo you can also use the Neuron Persistent Cache feature that is enabled now as default in transformers-neuronx in SDK release 2.13: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/libraries/transformers-neuronx/transformers-neuronx-developer-guide.html#neuron-persistent-cache

**Comment by dacorvo** [comment]
Solved in AWS Neuron SDK 2.15.
